move all variables in sampleAndWaterMarkHistograms::innerSet

MikeSpreitzer opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->
/kind bug
/kind regression

**What this PR does / why we need it**:
This is a refactoring to tiptoe around https://github.com/golang/go/issues/43570 for #97685

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #97685 

**Special notes for your reviewer**:
This is an alternative to #97791 and #97854

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Fixes a performance regression in 1.19 on ARM architectures computing priority and fairness metrics
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

MikeSpreitzer committed:
move all variables in sampleAndWaterMarkHistograms::innerSet

to tiptoe around https://github.com/golang/go/issues/43570 for #97685

MikeSpreitzer commented:
@kubernetes/sig-api-machinery-bugs

MikeSpreitzer commented:
@liggitt 
@erikwilson 
@brandond 
@tkashem 
@adtac

liggitt commented:
/lgtm
/approve

liggitt commented:
/hold for verification on ARM

brandond commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/sample_and_watermark.go:
```diff
@@ -160,17 +160,14 @@ func (saw *sampleAndWaterMarkHistograms) SetX1(x1 float64) {
-	var acc sampleAndWaterMarkAccumulator
-	var wellOrdered bool
-	func() {
+	when, whenInt, acc, wellOrdered := func() (time.Time, int64, sampleAndWaterMarkAccumulator, bool) {
```
At this point it seems like there's not much point in keeping it as an anonymous function, might as well just move that whole block out to `func (saw *sampleAndWaterMarkHistograms) validateQuantization(time.Time, int64, sampleAndWaterMarkAccumulator, bool)` (name it something better) unless the intent is to revert all of this once the golang bug is fixed.

liggitt replied:
the point of the inner func is to unlock via a `defer` as soon as our calculations are done before we log or call out to record metrics

brandond replied:
The side effects argument makes sense I suppose. I get that it just exists to handle the unlock cleanly, but to me the code picks up a bit of an odor when anonymous functions grow return values and such. I'm just here as peanut gallery though, feel free to ignore my feedback.

MikeSpreitzer replied:
I agree with @liggitt , the inner function should never be invoked in any other context and so I would prefer to not expose it.

MikeSpreitzer commented:
This fixes the problem for 386 on linux.

```
mspreitz@ubu20:~/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics$ git checkout test-saw
Switched to branch 'test-saw'
Your branch is up to date with 'origin/test-saw'.
mspreitz@ubu20:~/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics$ git branch -D test-and-fix
Deleted branch test-and-fix (was 91ef7229c36).
mspreitz@ubu20:~/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics$ git pull
remote: Enumerating objects: 33, done.
remote: Counting objects: 100% (33/33), done.
remote: Compressing objects: 100% (6/6), done.
remote: Total 33 (delta 27), reused 33 (delta 27), pack-reused 0
Unpacking objects: 100% (33/33), 3.00 KiB | 439.00 KiB/s, done.
From https://github.com/MikeSpreitzer/kubernetes
 + 714f40f3be9...14f94b6c5e5 rejigger-quantize-3  -> origin/rejigger-quantize-3  (forced update)
 * [new branch]              rejigger-quantize-3a -> origin/rejigger-quantize-3a
 * [new branch]              rejigger-quantize-4  -> origin/rejigger-quantize-4
Already up to date.
mspreitz@ubu20:~/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics$ git checkout -b test-and-fix
Switched to a new branch 'test-and-fix'
mspreitz@ubu20:~/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics$ git cherry-pick origin/rejigger-quantize-4
[test-and-fix d2613d7a69f] move all variables in sampleAndWaterMarkHistograms::innerSet
 Date: Fri Jan 8 13:32:38 2021 -0500
 1 file changed, 7 insertions(+), 9 deletions(-)
mspreitz@ubu20:~/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics$ 
mspreitz@ubu20:~/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics$ 
mspreitz@ubu20:~/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics$ GOARCH=386 go test -c .
mspreitz@ubu20:~/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics$ for i in {1..100}; do if GOGC=1 ./metrics.test >/dev/null 2>&1  ; then echo -n . ; else echo -n "F" ; fi; done; echo ""
....................................................................................................
mspreitz@ubu20:~/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics$ for i in {1..100}; do if ./metrics.test >/dev/null 2>&1  ; then echo -n . ; else echo -n "F" ; fi; done; echo ""
....................................................................................................
```

liggitt commented:
can `/hold cancel` once this is verified on ARM

brandond commented:
I can verify on arm this afternoon, if nobody else beats me to it.

liggitt commented:
@MikeSpreitzer if you want to go ahead and open the cherry-pick to release-1.20 so CI is happy, that would help meet the 1.20.2 deadline of EOD

hasheddan reviewed (approved):
/lgtm

@MikeSpreitzer is this ready to go?

brandond commented:
Working for me on arm:

```
sysadm@pi02:kubernetes$ git log --pretty=oneline | head -n 3
ba7a0759472012798fc3862bcf1b63558f6df2c0 Explain warnings in new test
737d4e2cef29d1f2089452be36bf9d46423ef553 Add unit test for sample-and-watermark histograms
611184aa59d0cd40466bc3bc4b40a3712a038171 move all variables in sampleAndWaterMarkHistograms::innerSet
sysadm@pi02:kubernetes$ cd staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics
sysadm@pi02:metrics$ GOARCH=arm GOARM=7 go test -c .
sysadm@pi02:metrics$ for i in {1..100}; do if GOGC=1 ./metrics.test >/dev/null 2>&1  ; then echo -n . ; else echo -n "F" ; fi; done; echo ""
....................................................................................................
sysadm@pi02:metrics$ for i in {1..100}; do if ./metrics.test >/dev/null 2>&1  ; then echo -n . ; else echo -n "F" ; fi; done; echo ""
....................................................................................................
```

liggitt commented:
thanks

/hold cancel

caesarxuchao commented:
/triage accepted
